test/CMakeLists.txt

```diff
@@ -1,5 +1,9 @@
 include(HalideTestHelpers)
 
+if (TARGET_SPIRV)
+    add_compile_definitions(WITH_SPIRV)
+endif ()
+
```

Comment:
We avoid directory commands like `add_compile_definitions` in this codebase. This define should probably be attached to the `Halide` library via `target_compile_definitions(Halide INTERFACE WITH_SPIRV)`

Reply:
So, the issue I faced here was the precompiled headers don't grab target specific compile definitions, only project level compile definitions.  Also, `$<BUILD_INTERFACE:>` doesn't appear to pass the definitions down for compiling object files.

Reply:
I worked around the precompiled headers by always including a stub function for the internal test.

Reply:
> the precompiled headers don't grab target specific compile definitions, only project level compile definitions

I don't think that's _quite_ true. They grab the target-specific compile definitions for `_test_internal` and then the same header is reused everywhere. See `test/CMakeLists.txt` and `cmake/HalideTestHelpers.cmake` for the relevant uses of `target_precompile_headers`.

Reply:
Hrmm ... I'm not entirely sure why, but the precompiled headers for Halide.h resulted in an error saying that they were invalid if I defined a target_compile_definitions(_internal_test PRIVATE WITH_SPIRV), even though it was also defined in the Halide compile definitions.

Reply:
You should look at the verbose output from a Ninja build to diagnose this. I wrote the following sample build to probe the behavior of `target_precompile_headers` and what I said was correct re: which flags it sees.

```
cmake_minimum_required(VERSION 3.23)
project(test)

string(APPEND CMAKE_CXX_FLAGS " -DGLOBAL")

add_compile_definitions(DIRECTORY)

add_library(impl impl.cpp)
target_compile_definitions(impl PUBLIC $<BUILD_INTERFACE:PROPAGATED>)
target_precompile_headers(impl PRIVATE impl.hpp)

add_executable(app main.cpp)
target_compile_definitions(app PRIVATE LOCAL)
target_link_libraries(app PRIVATE impl)
target_precompile_headers(app REUSE_FROM impl)
```
Create empty files:
```
$ touch main.cpp impl.{hpp,cpp}
```
Configure the build with Ninja:
```
$ cmake -G Ninja -S . -B build 
```
Run a dry-run:
```
$ cmake --build build/ -- -nv
[1/5] /usr/bin/c++ -DDIRECTORY -DPROPAGATED  -DGLOBAL -Winvalid-pch -x c++-header -include /home/reinking/test/build/CMakeFiles/impl.dir/cmake_pch.hxx -MD -MT CMakeFiles/impl.dir/cmake_pch.hxx.gch -MF CMakeFiles/impl.dir/cmake_pch.hxx.gch.d -o CMakeFiles/impl.dir/cmake_pch.hxx.gch -c /home/reinking/test/build/CMakeFiles/impl.dir/cmake_pch.hxx.cxx
[2/5] /usr/bin/c++ -DDIRECTORY -DPROPAGATED  -DGLOBAL -Winvalid-pch -include /home/reinking/test/build/CMakeFiles/impl.dir/cmake_pch.hxx -MD -MT CMakeFiles/impl.dir/impl.cpp.o -MF CMakeFiles/impl.dir/impl.cpp.o.d -o CMakeFiles/impl.dir/impl.cpp.o -c /home/reinking/test/impl.cpp
[3/5] /usr/bin/c++ -DDIRECTORY -DLOCAL -DPROPAGATED  -DGLOBAL -Winvalid-pch -include /home/reinking/test/build/CMakeFiles/impl.dir/cmake_pch.hxx -MD -MT CMakeFiles/app.dir/main.cpp.o -MF CMakeFiles/app.dir/main.cpp.o.d -o CMakeFiles/app.dir/main.cpp.o -c /home/reinking/test/main.cpp
[4/5] : && /usr/bin/cmake -E rm -f libimpl.a && /usr/bin/ar qc libimpl.a  CMakeFiles/impl.dir/impl.cpp.o && /usr/bin/ranlib libimpl.a && :
[5/5] : && /usr/bin/c++ -DGLOBAL  CMakeFiles/app.dir/main.cpp.o -o app  libimpl.a && :
```